### <p style="text-align:center;"> BATTLE OF THE NEIGHBOURHOODS - CANADA </p>

> **Name:** Jeff Mac Osei<br>
**Course:** Applied Data Science Capstone<br>
**Date:**25/04/2021

This is the first peer graded assignment for the applied data science capstone course.

In [1]:
import pandas as pd
import numpy as np
import requests

In [2]:
# Getting access to the wiki page to scrape data

wiki_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page_handle = requests.get(wiki_url)

In [3]:
# having a glimpse of the page

page_data = pd.read_html(page_handle.text)
page_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [4]:
# inspecting the size of our data and selecting the right table
len(page_data)

3

In [5]:
page_data = page_data[0]
page_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

#### Question 1

In [6]:
main_data = page_data[page_data['Borough'] != 'Not assigned']

In [7]:
main_data.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [8]:
main_data['Postal Code'].duplicated().sum()

0

In [9]:
main_data[page_data['Neighbourhood'] == 'Not assigned']

<ipython-input-9-91deecfee51c>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  main_data[page_data['Neighbourhood'] == 'Not assigned']


Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

> We can see that the `Postal Code` column has no duplicate values and the `Neighbourhood` column also has all assigned values

In [10]:
main_data.shape

(103, 3)

#### Question 2

In [11]:
import geocoder

In [12]:
geo_data = pd.read_csv('https://cocl.us/Geospatial_data')

In [13]:
geo_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [14]:
# checking the shape of the dataframe
geo_data.shape

(103, 3)

In [15]:
# combining two data set to form one dataframe

main_geo_data = main_data.join(geo_data.set_index('Postal Code'), on='Postal Code', how='inner')

In [16]:
main_geo_data.head()

Postal Code           Borough                                Neighbourhood  \
2         M3A        North York                                    Parkwoods   
3         M4A        North York                             Victoria Village   
4         M5A  Downtown Toronto                    Regent Park, Harbourfront   
5         M6A        North York             Lawrence Manor, Lawrence Heights   
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
2  43.753259 -79.329656  
3  43.725882 -79.315572  
4  43.654260 -79.360636  
5  43.718518 -79.464763  
6  43.662301 -79.389494

In [17]:
# check for shape of new dataframe
main_geo_data = main_geo_data.reset_index()

In [18]:
main_geo_data.drop(['index'], axis = 'columns', inplace = True)

In [19]:
main_geo_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

#### Question 3

We will segment and cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [20]:
from geopy.geocoders import Nominatim 
import folium

In [21]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [23]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(main_geo_data['Latitude'], main_geo_data['Longitude'], main_geo_data['Borough'], main_geo_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [24]:
CLIENT_ID = 'SEEWRB4CKEUA20RKFEUGG1ZGZ0R5NSQJ2451B3QMARJJUQ1U'
CLIENT_SECRET = 'DBXUP2G14FPLIC4ZTT3RBHFVXJKRZVK5J2FUAMFDQ5QEJPM4'
VERSION = '20201215' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: SEEWRB4CKEUA20RKFEUGG1ZGZ0R5NSQJ2451B3QMARJJUQ1U
CLIENT_SECRET:DBXUP2G14FPLIC4ZTT3RBHFVXJKRZVK5J2FUAMFDQ5QEJPM4


Next, we create a function to get all the venue categories in Toronto

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [27]:
venues_in_toronto = getNearbyVenues(main_geo_data['Neighbourhood'], main_geo_data['Latitude'], main_geo_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [28]:
venues_in_toronto.shape

(1322, 5)

> So we have 1322 records and 5 columns. Checking sample data

In [29]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood

In [30]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   
1310  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572             Tim Hortons   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1306               43.628841               -79.520999        South St. Burger   
1307               43.628841               -79.520999              Wingporium   
1308               43.628841               -79.520999               Dollarama   
1309               43.628841               -79.520999    Artisano Bakery Café   
1310               43.628841               -79.520999          Healthy Planet   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3               Coffee Shop  
4     Portuguese Restaurant  
...                     ...  
1306           Burger Joint  
1307            Wings Joint  
1308         Discount Store  
1309                 Bakery  
1310        Supplement Shop  

[414 rows x 5 columns]

> So there are 414 records for each neighbourhood.
Checking for the maximum venue categories

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.718518               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[237 rows x 4 columns]

There are around 237 different types of Venue Categories.

## One Hot encoding the venue Categories

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1317                  0        0                   0             0   
1318                  0        0                   0             0   
1319                  0        0                   0             0   
1320                  0        0                   0             0   
1321                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1317               0                0                 0                    0   
1318               0                0                 0                    0   
1319               0                0                 0                    0   
1320               0                0                 0                    0   
1321               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  Turkish Restaurant  \
0                0         0  ...              0                   0   
1                0         0  ...              0                   0   
2                0         0  ...              0                   0   
3                0         0  ...              0                   0   
4                0         0  ...              0                   0   
...            ...       ...  ...            ...                 ...   
1317             0         0  ...              0                   0   
1318             0         0  ...              0                   0   
1319             0         0  ...              0                   0   
1320             0         0  ...              0                   0   
1321             0         0  ...              0                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1317                              0                 0                      0   
1318                              0                 0                      0   
1319                              0                 0                      0   
1320                              0                 0                      0   
1321                              0                 0                      0   

      Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                   0         0            0              0            0  
1                   0         0            0              0            0  
2                   0         0            0              0            0  
3                   0         0            0              0            0  
4  

Adding the neighbourhood to the encoded dataframe

In [33]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Turkish Restaurant  \
0                    0             0  ...              0                   0   
1                    0             0  ...              0                   0   
2                    0             0  ...              0                   0   
3                    0             0  ...              0                   0   
4                    0             0  ...              0                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0                0         0            0              0            0  
1                0         0            0              0            0  
2                0         0            0              0            0  
3                0         0            0              0            0  
4                0         0            0              0            0  

[5 rows x 238 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.045455           0.0  ...            0.0   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 238 columns]

Let's make a function to get the top most common venue categories

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1             Sandwich Place              Pharmacy           Coffee Shop   
2                       Bank              Pharmacy    Chinese Restaurant   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4             Sandwich Place    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store          Dessert Shop     Electronics Store   
1                   Pub                   Gym       Airport Service   
2         Shopping Mall                 Diner        Sandwich Place   
3           Yoga Studio          Dessert Shop     Electronics Store   
4             Juice Bar               Butcher            Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0  Eastern European Restaurant          Dumpling Restaurant   
1                  Escape Room  Eastern European Restaurant   
2                Deli / Bodega                   Restaurant   
3  Eastern European Restaurant          Dumpling Restaurant   
4                         Café            Indian Restaurant   

  10th Most Common Venue  
0              Drugstore  
1    Dumpling Restaurant  
2            Supermarket  
3              Drugstore  
4       Sushi Restaurant

Let's make the model to cluster our Neighbourhoods

In [37]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [38]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [39]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 4, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1,
       0, 1, 1, 1, 1, 1, 0, 2])

Let's add the clustering Label column to the top 10 common venue categories

In [40]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [42]:
toronto_merged = main_geo_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio      Department Store   
1           Coffee Shop          Intersection          Hockey Arena   
2                  Park                Bakery                   Pub   
3     Accessories Store              Boutique             Gift Shop   
4           Yoga Studio                 Diner        Sandwich Place   

    5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0       Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
1   Portuguese Restaurant            French Restaurant            Comic Shop   
2          Breakfast Spot               Chocolate Shop   Distribution Center   
3  Furniture / Home Store                  Event Space           Coffee Shop   
4   Portuguese Restaurant                         Park    Mexican Restaurant   

     8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                Drugstore             Donut Shop                Dog Run  
1  Comfort Food Restaurant    Dumpling Restaurant              Drugstore  
2                      Spa           Dessert Shop            Event Space  
3            Women's Store  Vietnamese Restaurant       Airport Terminal  
4       Italian Restaurant             Hobby Shop    Fried Chicken Joint

Drop all the NaN values to prevent data skew

In [43]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [44]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [45]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
35         East York             0.0                  Park   
49        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio      Department Store   
21               Women's Store                  Pool           Yoga Studio   
35           Convenience Store          Intersection           Yoga Studio   
49  Construction & Landscaping                Bakery           Yoga Studio   
61                 Swim School              Bus Line           Yoga Studio   
64                 Yoga Studio         Deli / Bodega     Electronics Store   
66           Convenience Store           Yoga Studio      Department Store   
85                        Park                Bakery          Intersection   
91                       Trail            Playground           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21                Deli / Bodega  Eastern European Restaurant   
35             Department Store            Electronics Store   
49             Department Store            Electronics Store   
61             Department Store  Eastern European Restaurant   
64  Eastern European Restaurant          Dumpling Restaurant   
66            Electronics Store  Eastern European Restaurant   
85                  Yoga Studio             Department Store   
91             Department Store                 Dessert Shop   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Dumpling Restaurant             Drugstore            Donut Shop   
21          Dumpling Restaurant             Drugstore            Donut Shop   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
49  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
61          Dumpling Restaurant             Drugstore            Donut Shop   
64                    Drugstore            Donut Shop               Dog Run   
66          Dumpling Restaurant             Drugstore            Donut Shop   
85  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
91           Dim Sum Restaurant                 Diner        Discount Store   

   10th Most Common Venue  
0                 Dog Run  
21                Dog Run  
35             Donut Shop  
49             Donut Shop  
61                Dog Run  
64    Distribution Center  
66                Dog Run  
85             Donut Shop  
91                Dog Run

Cluster 2

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York             1.0                 Pizza Place   
2    Downtown Toronto             1.0                 Coffee Shop   
3          North York             1.0              Clothing Store   
4    Downtown Toronto             1.0                 Coffee Shop   
7          North York             1.0                         Gym   
..                ...             ...                         ...   
98          Etobicoke             1.0                        Pool   
99   Downtown Toronto             1.0                         Gym   
100      East Toronto             1.0          Light Rail Station   
101         Etobicoke             1.0  Construction & Landscaping   
102         Etobicoke             1.0                         Gym   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1               Coffee Shop          Intersection          Hockey Arena   
2                      Park                Bakery                   Pub   
3         Accessories Store              Boutique             Gift Shop   
4               Yoga Studio                 Diner        Sandwich Place   
7                Beer Store           Coffee Shop   Japanese Restaurant   
..                      ...                   ...                   ...   
98                    River           Yoga Studio          Dance Studio   
99                 Creperie        Breakfast Spot       Bubble Tea Shop   
100           Auto Workshop            Comic Shop           Pizza Place   
101          Baseball Field           Yoga Studio          Dessert Shop   
102  Thrift / Vintage Store     Convenience Store        Discount Store   

           5th Most Common Venue 6th Most Common Venue  \
1          Portuguese Restaurant     French Restaurant   
2                 Breakfast Spot        Chocolate Shop   
3         Furniture / Home Store           Event Space   
4          Portuguese Restaurant                  Park   
7                           Café    Dim Sum Restaurant   
..                           ...                   ...   
98   Eastern European Restaurant   Dumpling Restaurant   
99                  Burger Joint    Salon / Barbershop   
100             Recording Studio            Restaurant   
101                  Escape Room     Electronics Store   
102                Burrito Place          Burger Joint   

           7th Most Common Venue    8th Most Common Venue  \
1                     Comic Shop  Comfort Food Restaurant   
2            Distribution Center                      Spa   
3                    Coffee Shop            Women's Store   
4             Mexican Restaurant       Italian Restaurant   
7                    Supermarket               Restaurant   
..                           ...                      ...   
98                     Drugstore               Donut Shop   
99                    Restaurant                     Café   
100                      Butcher            Burrito Place   
101  Eastern European Restaurant      Dumpling Restaurant   
102               Sandwich Place              Social Club   

     9th Most Common Venue 10th Most Common Venue  
1      Dumpling Restaurant              Drugstore  
2             Dessert Shop            Event Space  
3    Vietnamese Restaurant       Airport Terminal  
4               Hobby Shop    Fried Chicken Joint  
7     Caribbean Restaurant     Chinese Restaurant  
..                     ...                    ...  
98                 Dog Run    Distribution Center  
99        Ramen Restaurant                    Pub  
100             Skate Park                Brewery  
101              Drugstore             Donut Shop  
102   Fast Food Restaurant                 Bakery  

[88 rows x 12 columns]

Cluster 3

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0   Martial Arts School           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
45      Department Store     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
45               Dog Run    Distribution Center

Cluster 4

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             3.0            Print Shop           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

Cluster 5

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             4.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6   Dumpling Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

We have successfully cluster Toronto neighbourhood based on venue categories!